File: solutions/java/code-carl/ch09_dp/T9_LC0416_CanPartition.java
/**
 * Practice template for:
 * - LeetCode 416. Partition Equal Subset Sum
 *   https://leetcode.cn/problems/partition-equal-subset-sum/
 */
/**
 * @author Cheng Yihao
 * @version 1.0
 * @date 2024/11/21 19:18
 * @comment
 */
public class T9_LC0416_CanPartition {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % 2 == 1) {
            return false;
        }

        int bagSize = sum / 2;
        int[] dp = new int[bagSize + 1];
        for (int i = 0; i < nums.length; i++) {
            for (int j = bagSize; j >= nums[i]; j--) {
                dp[j] = Math.max(dp[j], dp[j - nums[i]] + nums[i]);
            }
        }
        return dp[bagSize] == bagSize;
    }
}

File: templates/code-carl/ch09_dp/t9_lc0416_can_partition.py
"""LeetCode 416. Partition Equal Subset Sum
https://leetcode.cn/problems/partition-equal-subset-sum/
"""

from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.canPartition([1, 5, 11, 5]) == True
    assert solution.canPartition([1, 2, 3, 5]) == False
    print("All tests passed.")
